core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala

```diff
@@ -283,8 +312,55 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
     TestUtils.createTopic(zkUtils, topic, 1, brokerCount, servers, new Properties())
     servers.foreach { server =>
       val logDir = server.logManager.getLog(tp).get.dir.getParent
-      assertEquals(replicaAssignment(new TopicPartitionReplica(topic, 0, server.config.brokerId)), logDir)
+      assertEquals(firstReplicaAssignment(new TopicPartitionReplica(topic, 0, server.config.brokerId)), logDir)
     }
+
+    // Verify that replica can be moved to the specified log directory after the topic has been created
+    adminClient.alterReplicaDir(secondReplicaAssignment.asJava, new AlterReplicaDirOptions()).all().get()
+    servers.foreach { server =>
+      TestUtils.waitUntilTrue(() => {
+        val logDir = server.logManager.getLog(tp).get.dir.getParent
+        secondReplicaAssignment(new TopicPartitionReplica(topic, 0, server.config.brokerId)) == logDir
+      }, "timed out waiting for replica movement", 6000L)
+    }
+
+    // Verify that replica can be moved to the specified log directory while the producer is sending messages
+    val running = new AtomicBoolean(true)
+    @volatile var numMessages = 0
+    val thread = new Thread() {
+      override def run(): Unit = {
+        val producer = TestUtils.createNewProducer(
+          TestUtils.getBrokerListStrFromServers(servers),
+          retries = 5,
+          requestTimeoutMs = 2000,
+          acks = -1
+        )
+
+        while (running.get()) {
+          producer.send(new ProducerRecord(topic, s"xxxxxxxxxxxxxxxxxxxx".getBytes))
+          numMessages += 1
+        }
+        producer.close()
+      }
+    }
+
+    try {
+      thread.start()
+      Thread.sleep(3000)
```

Comment:
Why is this necessary? A comment to explain why would be useful.

Reply:
@tombentley Thanks for the comment. The patch you saw was fully ready for review. The patch has been updated now with better code.